client-lite/src/util/do_curl_multi_operation.cpp

```diff
@@ -0,0 +1,247 @@
+#include "do_common.h"
+#include "do_curl_multi_operation.h"
+
+#include <algorithm>
+
+CurlMultiOperation::CurlMultiOperation()
+{
+    _multiHandle = curl_multi_init();
+    THROW_HR_IF(E_OUTOFMEMORY, _multiHandle == nullptr);
+
+    _fKeepRunning = true;
+    _multiPerformThread = std::thread{[this]()
+        {
+            _WorkerThread();
+        }};
+}
+
+CurlMultiOperation::~CurlMultiOperation()
+{
+    {
+        std::unique_lock<std::mutex> lock{_mutex};
+        _fKeepRunning = false;
+        _cv.notify_one();
+    }
+    _multiPerformThread.join();
+
+    {
+        _activeHandles.RemoveAll(_multiHandle);
+        _handlesToAdd.clear();
+        _handlesToRemove.clear();
+    }
+
+    curl_multi_cleanup(_multiHandle);
+}
+
+void CurlMultiOperation::AddHandle(CURL* easyHandle, completion_callback_t pCallback, void* pCallbackUserData)
+{
+    std::unique_lock<std::mutex> lock{_mutex};
+    THROW_HR_IF(E_NOT_VALID_STATE, !_fKeepRunning);
+    auto itAdd = std::find(std::begin(_handlesToAdd), std::end(_handlesToAdd), easyHandle);
+    THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_EXISTS), itAdd != _handlesToAdd.end());
+    _handlesToAdd.emplace_back(HandleData{easyHandle, pCallback, pCallbackUserData});
+    _cv.notify_one();
+}
+
+void CurlMultiOperation::RemoveHandle(CURL* easyHandle)
+{
+    std::shared_ptr<WrappedHandleData> wrappedHandleToBeDeleted;
+    {
+        std::unique_lock<std::mutex> lock{_mutex};
+
+        _RemoveHandle(easyHandle, _handlesToAdd);
+
+        auto pExistingWrappedHandle = _activeHandles.Get(easyHandle);
+        if (pExistingWrappedHandle == nullptr)
+        {
+            return;
+        }
+        wrappedHandleToBeDeleted = *pExistingWrappedHandle;
+
+        if (std::find(std::begin(_handlesToRemove), std::end(_handlesToRemove), easyHandle) == _handlesToRemove.end())
+        {
+            _handlesToRemove.emplace_back(easyHandle);
+        }
+        _cv.notify_one();
+    }
+
+    if (wrappedHandleToBeDeleted)
+    {
+        // outside the lock, wait for the handle to be removed from _activeHandles
+        (void)wrappedHandleToBeDeleted->inactiveSignal.Wait(std::chrono::minutes(30));
+    }
+}
+
+void CurlMultiOperation::_WorkerThread()
+{
+    while (true)
+    {
+        std::unique_lock<std::mutex> lock{_mutex};
+        if (!_fKeepRunning)
+        {
+            break;
+        }
+
+        for (const auto& h : _handlesToAdd)
+        {
+            _activeHandles.Add(h, _multiHandle);
+        }
+        _handlesToAdd.clear();
+
+        for (const auto& h : _handlesToRemove)
+        {
+            _activeHandles.Remove(h, _multiHandle);
+        }
+        _handlesToRemove.clear();
+
+        if (_activeHandles.Empty())
+        {
+            _cv.wait(lock, [this](){ return !_fKeepRunning || !_handlesToAdd.empty(); });
```

Comment:
>!_handlesToAdd.empty() [](http://example.com/codeflow?start=62&length=22)

if this is non-empty, wouldn't cv wait here forever? There's no way to drain/clear the vector except for line 89 no?

Reply:
It is the opposite - wait() will return if _handlesToAdd becomes non-empty or _fKeepRunning becomes false.

Also note that AddHandle will wait if we were executing line 89-99 at that time (same mutex).